tpp-app/src/services/cycle/CycleService.js

```diff
@@ -0,0 +1,381 @@
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import {FLOW_LEVEL} from '../utils/constants';
+import {initializeEmptyYear, getDateString, getCalendarByYear, getSymptomsFromCalendar, getDaysDiffInclusive, getPeriodsInYear}  from '../utils/helpers';
+import {Symptoms} from '../utils/models';
+import differenceInCalendarDays from 'date-fns/differenceInDays';
+import isSameDay from 'date-fns/isSameDay';
+import subDays from 'date-fns/subDays';
+import addDays from 'date-fns/addDays';
+import Keys from '../utils/keys';
+import isAfter from 'date-fns/isAfter';
+
+
+
+
+
+/**
+ * Gets the end date of the final period in the year, which may be in the next year. This is not a prediction
+ * @param {Date} finalPeriodStart The start date of the final period of the year. Assumed to be a period day.
+ * @param {Object} calendar The object containing the symptoms for this year, last year, and next year. Optional.
+ * @return {Promise} Resolves into Date object that is the closest date after finalPeriodStart that is the end of a period. Date may be in a later year than finalPeriodStart
+ */
+async function getLastPeriodsEnd(finalPeriodStart, calendar = null){
+    var current = finalPeriodStart;
+    var yesterday = subDays(current, 1);
+    var twoDaysEarlier = subDays(current, 2);
+
+    if (!calendar){
+      calendar = await getCalendarByYear(current.getFullYear());
+    }
+
+    let dateSymptoms = await getSymptomsFromCalendar(calendar, current.getDate(), current.getMonth()+1, current.getFullYear());
+    let yesterdaySymptoms = new Symptoms();
+    let twoDaysEarlierSymptoms = new Symptoms()
+
+    var noFlowToday = (dateSymptoms.flow === FLOW_LEVEL.NONE || dateSymptoms.flow === null) ;
+    var noFlowYesterday = (yesterdaySymptoms.flow === FLOW_LEVEL.NONE || yesterdaySymptoms.flow === null) ;
+    var flowTwoDaysEarlier = (twoDaysEarlierSymptoms.flow !== null && twoDaysEarlierSymptoms.flow !== FLOW_LEVEL.NONE)
+
+
+    while(!(noFlowToday && noFlowYesterday && flowTwoDaysEarlier)){
```

Comment:
It seems like you are looking for _ _ X here. If you are looking for the end date of the final period in the next year shouldn't you be looking for X _ _? Or am I just confused 😅